src/Spago/Config.purs

```diff
@@ -174,130 +175,212 @@ type ReadWorkspaceOptions =
+        pure { workspace: { config: workspace, doc, rootPackage: package }, rootPath }
+      _ -> do
+        -- No workspace in this directory => recur to parent directory (unless it's already root)
+        when (parentDir == dir) $
+          dieForLackOfSpagoYaml
+        walkDirectoriesUpFrom parentDir
+
+    where
+    configFile = dir </> spagoYaml
+    parentDir = Path.dirname dir
+
+  loadSubprojectConfigs rootPath = do
+    candidates <- liftAff $ Glob.gitignoringGlob
+      { root: rootPath
+      , includePatterns: [ "**/" <> spagoYaml ]
+      , ignorePatterns: [ "**/node_modules/**", "**/.spago/**" ]
+      }
 
-  -- We prune any configs that use a different workspace.
-  -- For reasoning, see https://github.com/purescript/spago/issues/951
-  let configPathsWithWorkspaces = otherPackages # Array.mapMaybe \readResult -> readResult.packagePath <$ readResult.config.yaml.workspace
-  unless (Array.null configPathsWithWorkspaces) do
-    logDebug $ "Found these paths with workspaces: " <> String.joinWith ", " (Path.quote <$> configPathsWithWorkspaces)
+    -- Traversing directories (not files) and doing it in sorted order ensures
+    -- that parent directories come before their subdirectories. That way we
+    -- can remember workspaces that we find along the way and avoid trying to
+    -- load their subprojects that come later.
+    candidates <#> Path.toGlobal <#> Path.dirname # Array.sort # traverse_ \dir -> do
+      st <- State.get
+      let
+        configFile = dir </> spagoYaml
+        alreadyLoaded = st.loadedPackages # Map.member configFile
+        blockedSubtree = st.blockedSubtrees # Array.find (_ `Path.isPrefixOf` dir)
+      case alreadyLoaded, blockedSubtree of
+        true, _ ->
+          pure unit
+        _, Just b -> do
+          logDebug $ "Not trying to load " <> Path.quote configFile <> " because it is nested under " <> Path.quote b
+          pure unit
+        false, Nothing ->
+          readConfig' configFile >>= case _ of
+            Left _ -> do
+              logWarn [ "Failed to read config at " <> Path.quote configFile, "Skipping it and all its subprojects" ]
+              State.modify_ \s -> s { blockedSubtrees = Array.cons dir s.blockedSubtrees }
+            Right { yaml: { workspace: Just _ } } ->
+              State.modify_ \s -> s { blockedSubtrees = Array.cons dir s.blockedSubtrees }
+            Right config@{ yaml: { package: Just package } } -> do
+              logDebug $ "Loaded a subproject config at " <> Path.quote configFile
+              State.modify_ \s -> s { loadedPackages = Map.insert dir { package, config } s.loadedPackages }
+            Right _ -> do
+              logWarn $ "Neither workspace nor package found in " <> Path.quote configFile
+
+  tryReadConfigAt path = do
+    exists <- FS.exists path
+    if exists then
+      Just <$> do
+        logDebug $ "Loading spago.yaml at " <> Path.quote path
+        readConfig' path >>= rightOrDieWith \errLines ->
```

Comment:
This function is called in `walkDirectoriesUpFrom`, and the way I'm reading this is that a parse failure will terminate execution? Currently (in `readWorkspaceConfig`) we just log a warning and keep going, don't think we should be this harsh

Reply:
I don't think that's the right behavior. Would it really be ok to potentially operate on a whole different workspace just because we couldn't read one of the configs in this one?

Reply:
Also, this is a bit apples to oranges. In the previous version we do, in fact, error out if we can't read the root config: https://github.com/purescript/spago/blob/c723832d1474afac9efe6256a8e8feb4c85b567e/src/Spago/Config.purs#L198-L206

And only on nested configs we do the warning+ignore thing.

But here we don't know which one is the root one until we have read it successfully. So it's a different situation.

On the other hand, after we discover the root one and start reading back down the tree, we keep the previous behavior: warning+ignore. Except it was changed as a result of discussion in #1316 to warning+ignore the whole subtree.

Reply:
Ah I see, thanks for the explanation!